File: frontend/app/services/miembros-service.js
import Ember from "ember";
import ApiRequest from "msr-backoffice-frontend/utils/api-request";
import SingleToResourceService from "./single-to-resource-service";

/**
 * Servicio Miembros; injectado con el mixin MiembrosServiceInjected
 * Un miembro es el usuario del sistema que consume beneficios. El portador de una tarjeta
 **/
export default Ember.Service.extend(SingleToResourceService, {
  pathDelRecurso: 'miembros/',

  findAll(){
    return this.findWithFilter(Ember.Object.create({}));
  },

  findByNroTarjeta(nroTarjeta){
    return this.findWithFilter(Ember.Object.create({numeroDeTarjeta: nroTarjeta}));
  },

  findWithFilter(filtro){
    var apiRequest = ApiRequest.posting(this.pathDelRecurso, filtro);
    return this.sendAndEmberizeWithSpinner(apiRequest);
  },

  edit(id){
    var apiRequest = ApiRequest.getting(this.pathDelRecurso + id + '/edit');
    return this.sendAndEmberize(apiRequest);
  },

  altaTarjeta(data){
    var apiRequest = ApiRequest.posting(this.pathDelRecurso + data.get('memberId') + '/alta-tarjeta', data);
    return this.sendAndEmberize(apiRequest);
  },

  update(miembroEnEdicion){
    var apiRequest = ApiRequest.putting(this.pathDelRecurso + miembroEnEdicion.get('id'), miembroEnEdicion);
    return this.sendAndEmberize(apiRequest);
  },

  actualizarSaldosConValuelink(id){
    var apiRequest = ApiRequest.posting(this.pathDelRecurso + id + '/actualizar-saldos-con-vl');
    return this.sendAndEmberizeWithSpinner(apiRequest);
  },

  reporteDeBeneficiosAlDia(miembro){
    var apiRequest = ApiRequest.getting(`miembros/${miembro.id}/reporte-de-beneficios-al-dia/`);
    return this.sendAndEmberize(apiRequest);
  },

  informacionDeAuditoria(id){
    var apiRequest = ApiRequest.getting(`miembros/${id}/informacion-de-auditoria/`);
    return this.sendAndEmberize(apiRequest);
  },

  /**
   * Redefino los metodos heredados que todavía no deberíamos usar ni tenemos implementados en el backend
   */
  create(){
    throw new Error("Por ahora un miembro no puede ser creado");
  },

  deleteById(id, body) {
    let apiRequest = ApiRequest.posting("miembros/" + id + "/eliminar-miembro-rewards", body);
    return this.sendAndEmberizeWithSpinner(apiRequest);
  },
});

File: src/main/java/com/rewardsfrontend/service/MiembrosService.java
package com.rewardsfrontend.service;

import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import javax.servlet.http.HttpSession;
import java.util.Map;

@Service
public class MiembrosService extends BaseProxyService {

    public ResponseEntity<Map> buscarMiembros(Map<String, Object> filtro, HttpSession session) {
        return proxyPost("miembros", filtro, Map.class, session);
    }

    public ResponseEntity<Map> getMiembroParaEditar(Long id, HttpSession session) {
        return proxyGet("miembros/" + id + "/edit", Map.class, session);
    }

    public ResponseEntity<Map> actualizarMiembro(Long id, Map<String, Object> miembro, HttpSession session) {
        return proxyPut("miembros/" + id, miembro, Map.class, session);
    }

    public ResponseEntity<Map> altaTarjeta(Long id, Map<String, Object> data, HttpSession session) {
        return proxyPost("miembros/" + id + "/alta-tarjeta", data, Map.class, session);
    }

    public ResponseEntity<Map> actualizarSaldosConVL(Long id, HttpSession session) {
        return proxyPost("miembros/" + id + "/actualizar-saldos-con-vl", null, Map.class, session);
    }

    public ResponseEntity<Map> reporteDeBeneficios(Long id, HttpSession session) {
        return proxyGet("miembros/" + id + "/reporte-de-beneficios-al-dia", Map.class, session);
    }

    public ResponseEntity<Map> informacionDeAuditoria(Long id, HttpSession session) {
        return proxyGet("miembros/" + id + "/informacion-de-auditoria", Map.class, session);
    }

    public ResponseEntity<Map> eliminarMiembro(Long id, Map<String, Object> body, HttpSession session) {
        return proxyPost("miembros/" + id + "/eliminar-miembro-rewards", body, Map.class, session);
    }
}
